### PR TITLE
Basic PersonModal for dropoff and completed paths

### PR DESCRIPTION
**Pull request by neilkakkar**
## Changes

Basic PersonModals for dropped off and completed states.

Open question: Do we want links to be clickable? i.e. people going from Path A -> Path B. These aren't covered by the existing over options.

image

image

## How did you test this code?

Run locally, see UI looks fine. Click on Paths dropoff count, ensure persons modal opens with right data

### CODE REVIEW

File: frontend/src/scenes/paths/pathsLogic.ts
Comment (neilkakkar) [inline frontend/src/scenes/paths/pathsLogic.ts:179]: @liyiy  any ideas? I traced the code a bit but this didn't make much sense to me

**Comment by liyiy** [comment]
Can we include screenshots/videos in UI PRs?

**Comment by neilkakkar** [comment]
Yes 100% sorry 😅 .

I've been so used to cleaning up the default text that I saw "include screenshots if UI" and I said "yes, not UI, thx". 😂

File: frontend/src/scenes/trends/PersonModal.tsx
Comment (neilkakkar) [inline frontend/src/scenes/trends/PersonModal.tsx:56]: Update text for dropoff or completion

**Comment by neilkakkar** [comment]
> Remove cohort button as it doesn't work

File: frontend/src/scenes/paths/pathsLogic.ts
Comment (liyiy) [reply frontend/src/scenes/paths/pathsLogic.ts:179]: This was being used for persons modal user search, but not sure if that's something we support still lol

**Review by EDsCODE — COMMENTED** [review]
When you click the 0 completed at a last step, it still returns people in the modal. I think it's returning the dropoff persons

**Comment by neilkakkar** [comment]
huh, something's definitely off here. That Path shouldn't exist if 0 completed it in the last step? Checking why it's like that O.o - sounds like a calculation gotcha.

**Comment by neilkakkar** [comment]
oh, and for the last items, the calculation changes as well. Fixing.

**Comment by liyiy** [comment]
re: the open question, shouldn't this dropdown option open up the persons modal? The dropdown is what I'm currently working on, are we overlapping?

Screen Shot 2021-09-29 at 2 00 06 PM

**Comment by EDsCODE** [comment]
I think persons should be viewable on the numbers being clicked or the links being clicked. The options menu should be actions directly related to setting the startpoint/endpoint/exclusion/viewing funnels. We could consider adding a "view persons" button like shown in your picture which would show all persons at that step, but I think the granularity with persons based on click of dropoff/completion/per link is most useful right now

**Comment by liyiy** [comment]
@EDsCODE that makes sense! I'll just hold off on that dropdown option until this gets merged in ^_^

**Comment by neilkakkar** [comment]
Okay, there was a lot of confusion w.r.t what dropoff and completion means. I realised there's no obvious definition for either.

In the end, I've settled for:

At a node, Completion means: The number of people who **reached** this node.
At a node, Dropoff means: The number of people who **reached** this node, and **didn't continue further ahead**.

---

Q: Why not consider dropoff to mean the number of people who didn't reach this node? That makes sense in Funnels, why not here?
I considered this, but the problem is that there's multiple places people can go from a previous path. How do you decide that those people intended to come to this path beforehand, and didn't come?
If we choose all people whose last node was the previous node, then all items in the same column get the same dropoff, which feels wrong.

---

Implications: 

1. Dropoff will always be smaller than completed count (since dropoff is a subset of completion).
2. Dropoff doesn't make sense for last node, since it will always be equal to completion count.
3. Dropoff doesn't make sense for first node, since no edges
4. Explain in the UI what these counts mean



Do we want to change completion to mean "Reached this node and **continued ahead**? That felt wrong for a node, since you're then implying it's not completed even if it reached this path node. Maybe call it something other than Completed if we want to use this definition?

cc: @paolodamico & @clarkus for further thoughts

**Comment by neilkakkar** [comment]
The next question this raises is: What should the completion and dropoff % mean?

For dropoff: The `percentage = Dropoff / Completion` makes sense, since Completed is like the total, and Dropoff is a part of that. So this shows % who didn't continue.

For Completion %: Perhaps % of people who made it here from the previous step makes sense? Like, if 100 people started at A, and 50 made it to B, and 20 made it to C; then for the path A->B->C:

At B:
Completed = 50
Dropoff = 30

Completed % = 50% (calculated from A as base)
Dropoff % = 60% (calculated from B as base)

----

If instead we change Completed % to mean % who continued, then

At B:
Completed = 50
Dropoff = 30

Completed % = 40% (calculated from B as base)
Dropoff % = 60% (calculated from B as base)

It looks more consistent, but feels useless.

----

Edit: More complicated example: What if multiple previous nodes lead to B? I'd say B's base for completion should be sum of completion of previous nodes. So:

(90)A--30-->B--10-->C
(110)X--20-->B--20-->Z

In this case, at B:
Completed = 50
Dropoff = 50 - 10 - 20 = 20

Dropoff % = 20 / 50 = 40%
Completed % = (30+20) / (90+110) = 25%

File: frontend/src/scenes/paths/NewPaths.js
Comment (neilkakkar) [inline frontend/src/scenes/paths/NewPaths.js:120]: @liyiy was there any specific reason for justify?

Left is important (to me) because it keeps paths that end early to the left, instead of at the rightmost end. It messes up endpoint sankey quite a bit 😅

Comment (liyiy) [reply frontend/src/scenes/paths/NewPaths.js:120]: I think changing it back to left is fine as long as the cards look okay 👍 Think I was playing around with it originally because @EDsCODE and I discussed that it was super confusing for some node displays to be on both sides (that weren't the last one)

**Comment by EDsCODE** [comment]
RE: what the numbers mean

I'm trying to add another number in #6161 for this reason. I was thinking the number in #6161 will be the total number of people that are "at" the node (which means they made it to the node). Then the dropoff is the number of people who didn't go anywhere further than that node. The completed is the number of people who moved on from that node. Then the percentages will make sense because it will be completed/total at node and dropoff/total at node.

> Maybe call it something other than Completed if we want to use this definition?

I'd agree. I'd also agree this number feels less useful. But I think we should leave the related numbers (percentages from one node to another) to be shown when you hover edges and leave the numbers at a node specifically about people at that node

> For Completion %: Perhaps % of people who made it here from the previous step makes sense? Like, if 100 people started at A, and 50 made it to B, and 20 made it to C; then for the path A->B->C:

@liyiy and I talked about this originally and the numbers seemed confusing because it wouldn't be immediately obvious which numbers to relate with each other when glancing at the card

**Comment by clarkus** [comment]
> Implications:
>  - Dropoff will always be smaller than completed count (since dropoff is a subset of completion).
>  - Dropoff doesn't make sense for last node, since it will always be equal to completion count.
>  - Dropoff doesn't make sense for first node, since no edges
>  - Explain in the UI what these counts mean

This aligns 100% with how I've been thinking about it. Since we're solving for an exploratory case, we're showing top path items. For each path item, we summarize the metrics as you described. One thing to consider are sessions that start mid-path. What if we have a path (A > B > C > D), some persons could start into the flow at B or C. It might be represented the same as dropoffs, just as an flowing-in of persons instead of a dropping-off.

**Comment by neilkakkar** [comment]
>  One thing to consider are sessions that start mid-path. What if we have a path (A > B > C > D), some persons could start into the flow at B or C. It might be represented the same as dropoffs, just as an flowing-in of persons instead of a dropping-off.

This wouldn't happen, as sessions starting at B would show up in a separate flow ( a new row below A, starting with B). Effectively, the 2 Bs are different items.

**Comment by neilkakkar** [comment]
New updates!

Dropdown now looks like this: (Notice the wording change to present test to indicate better what's going on)
image

image

And all 3 numbers are clickable (on the card, and 2 in the dropdown) - and show their corresponding people.

**Comment by neilkakkar** [comment]
I'll leave polishing out the text to another PR, if needed. Limiting scope to: Person Modal working fine 🌟 .

A weirdness we need to iron out in the UI: Paths are per session, which means a path item showing count as 5 can have just 1 person on that path, if that person, say, did the same thing over 5 days.

**Comment by neilkakkar** [comment]
@posthog-contributions-bot is my code winning this argument?

**Review by EDsCODE — APPROVED** [review]

**Pull request merged**
